sys/dev/isp/ispvar.h

```diff
@@ -613,18 +661,24 @@ struct ispsoftc {
 #define	ISP_CODE_ORG			0x1000	/* default f/w code start */
 #define	ISP_CODE_ORG_2300		0x0800	/* ..except for 2300s */
 #define	ISP_CODE_ORG_2400		0x100000 /* ..and 2400s */
-#define	ISP_FW_REV(maj, min, mic)	((maj << 24) | (min << 16) | mic)
-#define	ISP_FW_MAJOR(code)		((code >> 24) & 0xff)
-#define	ISP_FW_MINOR(code)		((code >> 16) & 0xff)
-#define	ISP_FW_MICRO(code)		((code >>  8) & 0xff)
-#define	ISP_FW_REVX(xp)			((xp[0]<<24) | (xp[1] << 16) | xp[2])
+#define	ISP_FW_REV(major, minor, micro)				\
+	((major << 16) & 0xff0000 | (minor << 8) & 0x00ff00 | micro & 0x0000ff)
+#define	ISP_FW_MAJOR(code)		((code & 0xff0000) >> 16)
+#define	ISP_FW_MINOR(code)		((code & 0x00ff00) >> 8)
+#define	ISP_FW_MICRO(code)		(code & 0x0000ff)
+#define	ISP_FW_REVX(xp)						\
+	((xp[0] << 16) & 0xff0000 | (xp[1] << 8) & 0x00ff00 | xp[2] & 0x0000ff)
```

Comment:
While you are editing them, macro arguments should be in parentheses.  And would you apply masks and shifts in different order the strings could be shorter.  Not sure why did you change them.

Reply:
The old code was wrong for the micro version:
```
#define ISP_FW_REV(maj, min, mic)        ((maj << 24) | (min << 16) | mic)
#define ISP_FW_MICRO(code)               ((code >>  8) & 0xff)
```
Additionally, I don't think we will see a firmware version with major|minor|micro > 255 anytime soon. To represent 255.255.255 shifting like this `((maj << 16) | (min << 8) | mic)` is enough.

Here is a second try taking your notes into account:
```
#define ISP_FW_REV(maj, min, mic)       ((maj << 16) | (min << 8) | mic)
#define ISP_FW_MAJOR(code)              ((code >> 16) & 0xff)
#define ISP_FW_MINOR(code)              ((code >> 8) & 0xff)
#define ISP_FW_MICRO(code)              (code & 0xff)
#define ISP_FW_REVX(xp)                 ((xp[0] << 16) | (xp[1] << 8) | xp[2])
```
What do you think about it?

Reply:
To be safe it should be like this (see additional parentheses around `code`):
```
#define ISP_FW_MAJOR(code)              (((code) >> 16) & 0xff)
```

Reply:
So
```c
#define ISP_FW_REV(maj, min, mic)       (((maj) << 16) | ((min) << 8) | (mic))
#define ISP_FW_MAJOR(code)              (((code) >> 16) & 0xff)
#define ISP_FW_MINOR(code)              (((code) >> 8) & 0xff)
#define ISP_FW_MICRO(code)              ((code) & 0xff)
#define ISP_FW_REVX(xp)                 (((xp[0]) << 16) | ((xp[1]) << 8) | (xp[2]))
```
then?

Reply:
In the last case parentheses should be around `xp`, not `xp[0]`.

Reply:
Ok, got it now.